optima/calibration.py

```diff
@@ -224,14 +243,79 @@ def extractdata(xdata, ydata):
                         else: modelrow = modelrows[row]
                         datax, datay = extractdata(results.datayears, datarow) # Pull out the not-NaN values
+                        if doplot: rowname = 'total' if nrows==1 else pars['popkeys'][row]
                         for i,year in enumerate(datax): # Loop over each data point available
+                            count += 1
                             modelx = findinds(results.tvec, year) # Find the index of the corresponding time point
                             modely = modelrow[modelx] # Finally, extract the model result!
-                            thismismatch = abs(modely - datay[i]) / mean(datay+eps)
+                            if method=='mad': thismismatch = abs(modely - datay[i]) / mean(datay+eps)
+                            elif method=='mse': thismismatch = (modely - datay[i])**2
+                            elif method=='mape': thismismatch = abs(modely - datay[i]) / (datay[i]+eps)
+                            else:
+                                errormsg = 'autofit(): "method" incorrect; you entered "%s", but must be one of:\n' % method
+                                errormsg += '"mad"  = mean absolute deviation\n'
+                                errormsg += '"mse"  = mean squared error\n'
+                                errormsg += '"mape" = mean absolute percentage error'
+                                raise OptimaException(errormsg)
                             allmismatches.append(thismismatch)
                             mismatch += thismismatch
+                            
+                            if doplot:
+                                tmpdebugdata = (count, key, rowname, year, datay[i], modely, thismismatch, mismatch)
+                                printv('#%i. Key="%s" pop="%s" year=%f datay=%f modely=%f thismis=%f totmis=%f' % tmpdebugdata, 4, verbose)
+                                debugdata.append(tmpdebugdata)
+                            
+        
+        if doplot:
+            autofitresults['count'].append(len(autofitresults['count'])) #  Append new count
+            autofitresults['mismatch'].append(mismatch) #  Append mismatch
+            autofitresults['allmismatches'].append(array(allmismatches).flatten()) #  Append mismatch
+            
+            autofitfig.clear()
+            nplots = len(set([d[1]+d[2] for d in debugdata]))+2 # 1 is the mismatch
+            rows = ceil(sqrt(nplots))
+            cols = rows-1 if rows*(rows-1)>=nplots else rows
+            
+            subplot(rows,cols,1)
+            scatter(autofitresults['count'], autofitresults['mismatch'])
+            xlabel('Count')
+            ylabel('Mismatch')
+            ylim((0,ylim()[1]))
+            
+            subplot(rows,cols,2)
+            allmis = autofitresults['allmismatches'][-1] # Shorten
+            bar(arange(len(allmis)), allmis)
+            xlabel('Data point')
+            ylabel('Mismatch')
```

Comment:
that's true, that would be a fair bit of extra work though, and on the FE users just see 'improvement', so would be inclined to leave it for now...